### ISSUE TITLE
Orientation issue with "use message as screensaver"

### ISSUE THREAD

**Issue by ersi-dnd** [koreader/koreader#4098]
* KOReader version: v2015.11-1748-g6877ade_2018-06-25
* Device: Kobo H2O, Pocketbook Inkpad 2

#### Issue

Some orientation issue has appeared lately when screensaver used with the "use message as screensaver" setting. I will upload a pic to illustrate. As far as I understand, "use message as screensaver" should do the same as "as is" except with some message box. 

#### Steps to reproduce

1. Set the screensaver to "use message as screensaver"
2. Open a (pdf or djvu) file in landscape
3. Use the screenlock to see the screensaver

**Comment by ersi-dnd** [comment]
See the image how "use message as screensaver" distorts the screensaver. There's no such distortion with "as is". Tested with Kobo H2O edition 1 and Pocketbook Inkpad 2.

![20180723_122320 1](https://user-images.githubusercontent.com/6865273/43070247-060cc2a8-8e78-11e8-88f9-5bc30bbdea10.jpg)

**Comment by robert00s** [comment]
Strange, I cannot reproduce it on my Kindle and emulator. Can you try on epub file?

**Comment by ersi-dnd** [comment]
Not too strange that Kindle behaves differently. 

On my end (Kobo H2O and Pocketbook Inkpad 2) it affects even epubs, just tried. To repeat:
- Select "use message as screensaver" 
- Have the file in landscape mode
- use screenlock

**Comment by NiLuJe** [comment]
Huh. Can reproduce on a H2O, not on a PW2.

**Comment by poire-z** [comment]
Reproducible on GloHD too.
We switch to portrait before showing screensaver:
https://github.com/koreader/koreader/blob/60a6d5f542908a2a4a457fa05a868276bdd32572/frontend/device/generic/device.lua#L136-L145

It may look like  refresh/repain area / orientation bug, but I'm not sure it is.
That's done correctly I think: the message is shown in portrait, the book page is shown in portrait - but it is not re-rendered in portrait, it's still the landscape image that is shown.
(But if we were to do that correctly - switch the book to portrait to display it correctly in portrait - we'd need a re-render of the book, which could take >30s on big books ! we don't want that.)
So we'll always have a landscape mode image of the page to show on portrait, which can only be ugly.
I guess the quick and correct solution is for screensaver.lua to force the white background option when we are in landscape mode, so we don't display any of the book below (which is bad, or would take too long to render in portrait mode).

Or the above code to decide depending on screensaver options if the rotation is needed or not (which looks like more thinking and work :)

**Comment by poire-z** [comment]
It mignt need the same kind of trick for poweroff and reboot (which is less permanent if it is bad :).
https://github.com/koreader/koreader/blob/878834d400b3dec4e858fe323cbd60fd4a9d74c5/frontend/ui/uimanager.lua#L55-L69
(2 settings to check in each case: `poweroff_screensaver_type` and if it doesnt exist, fallback to the suspend one `screensaver_type`)

**Comment by NiLuJe** [comment]
@poire-z : Possibly, but that's already half-broken on my end anyway ;). (The message is weirdly aligned, stuck in the top-right corner with its borders off-screen).

**Comment by poire-z** [comment]
Something like the screenshots in #3619 (_Since I made these screenshots, I moved the message to the top right, so it does not hide the bottom widget parts_) ?
(May be I didn't test how it works when "message" is used :| )

**Comment by NiLuJe** [comment]
@poire-z : Exactly, yes, although I *feel* like it's a tiiiiny bit offscreen (the two sides of the border hitting the edge of the screen), I think?

Hard to be sure, since grabbing a snapshot is tricky given the timing ;).

And from a quick glance at the code, it handles message properly (at least for suspend), by overriding it ;).

**Comment by poire-z** [comment]
Mhhh, no comment says anything, and I don't see any trick in the code, but I may have wished them to be like that (find that prettier), and it worked as-is without any trick :)

I remember I wished, made, and added a comment :) for something like that in 
https://github.com/koreader/koreader/blob/a0e803bd78ac8090c35146f6553acfc888b81549/frontend/ui/widget/trapwidget.lua#L88-L90
![image](https://user-images.githubusercontent.com/24273478/43245499-c22ffbc0-90ae-11e8-83a4-7f7667b6451d.png)
(from the animated GIF in #3609, when holding on a wikipedia image)

**Comment by NiLuJe** [comment]
@poire-z: That's perfectly fine with me, I just found it surprising because I wasn't *expecting* to see it there ;).

**Comment by ersi-dnd** [comment]
On Kobo H2O Edition 1, this issue is still there, or maybe was reintroduced by regression. When "Screensaver/Use message as screensaver" is selected, the document is landscape, and the device goes to sleep, the view gets mangled as shown in the first image in the thread.

No problem when the setting is "Screensaver/Leave screen as it is".

Under Version, I have v2018.12.

Merry Christmas.

**Comment by ersi-dnd** [comment]
In fact, I had done a clean re-install of Koreader some three weeks earlier. I backed up and retained only `defaults.persistent.lua`and `history.lua`.

At the same time I also updated Nickel to 4.12 to see just for fun if it will continue working. And I think I have KFMon too, even though I have no idea what it does. 

I'll simply accept that it's only my personal Koreader whose screensaver misbehaves.
